lib/registry/src/publish.rs

```diff
@@ -0,0 +1,243 @@
+use crate::graphql::{execute_query_modifier_inner, get_signed_url, GetSignedUrl};
+use crate::graphql::{publish_package_mutation_chunked, PublishPackageMutationChunked};
+use crate::{format_graphql, PartialWapmConfig};
+use console::{style, Emoji};
+use graphql_client::GraphQLQuery;
+use indicatif::{ProgressBar, ProgressState, ProgressStyle};
+
+use std::collections::BTreeMap;
+use std::fmt::Write;
+use std::io::BufRead;
+use std::path::PathBuf;
+
+static UPLOAD: Emoji<'_, '_> = Emoji("⬆️  ", "");
+static PACKAGE: Emoji<'_, '_> = Emoji("📦  ", "");
+
+#[derive(Debug, Clone)]
+pub enum SignArchiveResult {
+    Ok {
+        public_key_id: String,
+        signature: String,
+    },
+    NoKeyRegistered,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn try_chunked_uploading(
+    registry: Option<String>,
+    token: Option<String>,
+    package: &wapm_toml::Package,
+    manifest_string: &String,
+    license_file: &Option<String>,
+    readme: &Option<String>,
+    archive_name: &String,
+    archive_path: &PathBuf,
+    maybe_signature_data: &SignArchiveResult,
+    archived_data_size: u64,
+    quiet: bool,
+) -> Result<(), anyhow::Error> {
+    let registry = match registry.as_ref() {
+        Some(s) => format_graphql(s),
+        None => {
+            #[cfg(not(test))]
+            let config = PartialWapmConfig::from_file();
+            #[cfg(test)]
+            let config = PartialWapmConfig::from_file("publish");
```

Comment:
What is this?

Reply:
This is a problem with cargo test: whenever I want to test `PartialWapmConfig::from_file` in an integration test, I have to give it the ID of the currently running test. Otherwise, I will either accidentally modify the global config.toml or (using a tempdir), subsequent calls to `PartialWapmConfig::from_file` cannot see the changes I made.

So for testing, I create a `config.test_1.toml`, `config.test_2.toml` file, etc. So that if I do:

```rust
#[test]  
fn test_1() {
    let completely_new_config_1 = PartialWapmConfig::from_file("test_1");
}

#[test]  
fn test_2() {
    let completely_new_config_2 = PartialWapmConfig::from_file("test_2");
}
```

... these test can run in parallel without interfering with each other (tempdir does not work as tempdir gets dropped at the end of the function call). I talked with Michael about this, if you have a better solution, I'm open for suggestions.

Reply:
This should be improved. Please chat with @Michael-F-Bryan or @theduke so they can provide ideas how to

Reply:
Two options:

* use an env var to override the config path
  That could run into issues though with concurrent tests, since env vars apply to the whole process
* Make the tested methods accept a config path/full config as an argument

Reply:
I would make `PartialWapmConfig::from_file()` always take a path that everything is calculated relative to (e.g. `wasmer_home: &Path`). That way, the `wasmer` CLI can pass in the `$WASMER_HOME` directory for loading the config, while a testing environment would use the `wasmer.toml` from a temporary directory.

My thinking is that nothing in `lib/registry/src/config.rs` should ever call `PartialWapmConfig::get_file_location()` or `PartialWapmConfig::get_folder()` because that essentially hard-codes the Wasmer home directory. I'd really like to avoid using `#[cfg(test)]` anywhere except in a `#[cfg(test)] mod tests { ... }`.

Reply:
Okay, I migrated to `wasmer_dir: Path` instead of using `test_name`